muxers/mplex/src/lib.rs

```diff
@@ -475,50 +475,51 @@ where C: AsyncRead + AsyncWrite
             endpoint: substream.endpoint,
         };
 
-        match poll_send(&mut inner, elem) {
-            Ok(Async::Ready(())) => Ok(to_write),
-            Ok(Async::NotReady) => Err(IoErrorKind::WouldBlock.into()),
-            Err(err) => Err(err),
+        match poll_send(&mut inner, elem)? {
+            Async::Ready(()) => Ok(Async::Ready(to_write)),
+            Async::NotReady => Ok(Async::NotReady)
         }
     }
 
-    fn flush_substream(&self, _substream: &mut Self::Substream) -> Result<(), IoError> {
+    fn flush_substream(&self, _substream: &mut Self::Substream) -> Poll<(), IoError> {
         let mut inner = self.inner.lock();
         let inner = &mut *inner; // Avoids borrow errors
 
-        match inner.inner.poll_flush_notify(&inner.notifier_write, 0) {
-            Ok(Async::Ready(())) => Ok(()),
-            Ok(Async::NotReady) => {
+        match inner.inner.poll_flush_notify(&inner.notifier_write, 0)? {
+            Async::Ready(()) => Ok(Async::Ready(())),
+            Async::NotReady => {
                 inner.notifier_write.to_notify.lock().insert(TASK_ID.with(|&t| t), task::current());
-                Err(IoErrorKind::WouldBlock.into())
-            },
-            Err(err) => Err(err),
+                Ok(Async::NotReady)
+            }
         }
     }
 
-    fn shutdown_substream(&self, substream: &mut Self::Substream) -> Poll<(), IoError> {
+    fn shutdown_substream(&self, sub: &mut Self::Substream, _: Shutdown) -> Poll<(), IoError> {
         let elem = codec::Elem::Reset {
-            substream_id: substream.num,
-            endpoint: substream.endpoint,
+            substream_id: sub.num,
+            endpoint: sub.endpoint,
         };
 
         let mut inner = self.inner.lock();
         poll_send(&mut inner, elem)
     }
 
     fn destroy_substream(&self, mut substream: Self::Substream) {
-        let _ = self.shutdown_substream(&mut substream);        // TODO: this doesn't necessarily send the close message
+        let _ = self.shutdown_substream(&mut substream, Shutdown::All);        // TODO: this doesn't necessarily send the close message
```

Comment:
I wonder if we can remove this, or if we don't we should remove the TODO.

Reply:
I think we can remove it.